Life-Pill/pharmacy-pos-main-backend
# Integrate Amazon S3 bucket for Image Storage

# Pull Request: Integrate Amazon S3 for Image Storage

## Title
Integrate Amazon S3 for Image Storage and Retrieval

## Description

This pull request introduces the integration of Amazon S3 for storing and retrieving images within the application. The changes encompass modifications to several controllers and service implementations to enable seamless interaction with the S3 bucket for managing various image-related operations.

### Changes Made

#### S3 Configuration:
- Added a new `S3Config` class to configure the Amazon S3 client using the provided AWS credentials and region.

#### Service Implementation:
- Updated the `S3ServiceIMPL` class to utilize the Amazon S3 client for uploading and retrieving images.
- Implemented methods for uploading, retrieving, and updating images for branches, employers, and items.

#### Controller Endpoints:
- Modified existing controller endpoints to support image-related operations using Amazon S3.
- Added new endpoints for uploading, retrieving, and updating images for branches, employers, and items.

### Endpoints Added/Modified:
- `POST /save-branch`: Save a branch with an image.
- `GET /view-branch-profile-image/{branchId}`: Retrieve the profile image of a branch.
- `PUT /update-branch-profile-image/{branchId}`: Update the profile image of a branch.
- `POST /save-employer-with-image`: Save an employer with an image.
- `GET /view-profile-image/{employerId}`: Retrieve the profile image of an employer.
- `PUT /update-employer-image/{employerId}`: Update the profile image of an employer.
- `POST /save-item-with-image`: Save an item with an image.
- `GET /view-item-image/{itemId}`: Retrieve the image of an item.
- `PUT /update-item-image/{itemId}`: Update the image of an item.

### Test Coverage:
- Ensured adequate test coverage for the new and modified endpoints to validate correct functionality.
- Implemented unit tests to verify the integration with Amazon S3 for image storage and retrieval.

## Test Plan

- **Unit Tests**: Execute the implemented unit tests for the modified and new controller endpoints using JUnit to ensure proper functionality.
- **Manual Testing**: Manually test each endpoint in a local environment to validate image upload, retrieval, and update operations.

### Test Scenarios:
1. Save a branch/employer/item with an image and verify successful upload.
2. Retrieve the profile image of a branch/employer/item and validate correctness.
3. Update the profile image of a branch/employer/item and confirm successful update.
4. Test error scenarios such as missing images or incorrect requests and ensure proper error handling.

## Checklist

- [x] Integrated Amazon S3 for image storage and retrieval.
- [x] Implemented controller endpoints for image-related operations.
- [x] Ensured comprehensive test coverage for all image-related functionality.
- [x] Manually tested image upload, retrieval, and update operations.

## Conclusion

This pull request enhances the application by integrating Amazon S3 for efficient storage and retrieval of images. Please review the changes and provide feedback or suggestions for further improvement.

Best regards,  
Pramitha Jayasooriya,
Backend Developer at LifePill,
https://pramithamj.me
